src/dense_layout.rs

```diff
@@ -0,0 +1,213 @@
+///     symmetric_coupling_map (bool): Is the coupling graph symmetric
+///     error_matrix (numpy.ndarray): A 2D array that represents the error
+///         rates on the target device. The diagonal is the measurement error
+///         error rate on that qubit and then the other positions are the
+///         2q/cx error rate between the 2 indices.
+///
+/// Returns:
+///     (rows, cols, best_map): A tuple of the rows, columns and the best
+///     mapping found by the function. This can be used to efficiently create
+///     a sparse matrix that maps the layout of virtual qubits
+///     (0 to ``num_qubits``) to the physical qubits on the coupling graph.
+#[pyfunction]
+pub fn best_subset(
+    py: Python,
+    num_qubits: usize,
+    coupling_adjacency: PyReadonlyArray2<f64>,
+    num_meas: usize,
+    num_cx: usize,
+    use_error: bool,
+    symmetric_coupling_map: bool,
+    error_matrix: PyReadonlyArray2<f64>,
+) -> PyResult<(PyObject, PyObject, PyObject)> {
+    let coupling_adj_mat = coupling_adjacency.as_array();
+    let coupling_shape = coupling_adj_mat.shape();
+    let err = error_matrix.as_array();
+    let avg_meas_err = err.diag().mean().unwrap();
+    let best_result = (0..coupling_shape[0])
+        .into_par_iter()
+        .map(|k| {
+            let mut subgraph: Vec<[usize; 2]> = Vec::with_capacity(num_qubits);
+            let bfs = bfs_sort(coupling_adj_mat, k);
+            let mut connection_count = 0;
+            for i in 0..num_qubits {
+                let node_idx = bfs[i];
+                for (node, j) in coupling_adj_mat
+                    .index_axis(Axis(0), node_idx)
+                    .iter()
+                    .enumerate()
+                {
+                    if *j == 0. {
+                        continue;
+                    }
+                    for counter in bfs.iter().take(num_qubits) {
+                        if node == *counter {
+                            connection_count += 1;
+                            subgraph.push([node_idx, node]);
+                            break;
+                        }
+                    }
+                }
+            }
+            let error = if use_error {
+                let mut ret_error = 0.;
+                let meas_avg = bfs
+                    .iter()
+                    .take(num_qubits)
+                    .map(|i| {
+                        let idx = *i;
+                        err[[idx, idx]]
+                    })
+                    .sum::<f64>()
+                    / num_qubits as f64;
+                let meas_diff = meas_avg - avg_meas_err;
+                if meas_diff > 0. {
+                    ret_error += num_meas as f64 * meas_diff;
+                }
+                let cx_sum: f64 = subgraph.iter().map(|edge| err[[edge[0], edge[1]]]).sum();
+                let mut cx_err = cx_sum / subgraph.len() as f64;
+                if symmetric_coupling_map {
+                    cx_err /= 2.;
```

Comment:
Does this imply that the `cx_sum` will be double what it actually should be if the user passed a symmetric coupling map? Or maybe that the subgraph will end up with half as many edges?

Reply:
Yeah, it implies it will be double, with a symmetric coupling map we'll have bidirectional cx gates so we'll have duplicate entries in the subgraph for each cx gate between any 2 qubits, this is just making it so we're only counting a single cx error rate between 2 qubits. That being said I did just copy this formula from the previous implementation. I think as part of #7113 I think we will probably have to refactor this part a bit and construct the error matrix differently.

Reply:
Sounds good. I'm wondering if `subgraph` itself would have twice as many items if this were a symmetric coupling map anyway, so I wonder if dividing by 2 is actually necessary (and if it's not, it'd make the result wrong).

Reply:
I think it'd be good to have a unit test to validate that `DenseLayout` works as expected wrt. error rates for symmetric coupling maps. Probably beyond the scope of the PR, since you haven't changed the logic.

Reply:
Yeah, I'll open an issue for it after this merges. The testing for this pass was a bit lacking so having more thorough testing would be a good thing in general.